File: 문제풀이/chap5-4/prob5-4-5.java
import java.util.*;

class Main {
	public static void main(String[] args) {
		// 입력
		Scanner sc = new Scanner(System.in);
		long N = sc.nextLong();
		int K = sc.nextInt();
		long[] V = new long[K + 1];
		for (int i = 1; i <= K; i++) {
			V[i] = sc.nextLong();
		}
		
		// 비트 전체 탐색
		long Answer = 0;
		for (int i = 1; i < (1 << K); i++) {
			int cnt = 0; // 선택된 숫자의 개수
			long lcm = 1; // 최소공배수
			for (int j = 0; j < K; j++) {
				if ((i & (1 << j)) != 0) {
					cnt += 1;
					lcm = LCM(lcm, V[j + 1]);
				}
			}
			long num = N / lcm; // 선택된 숫자 모두의 배수인 것의 개수
			if (cnt % 2 == 1) Answer += num;
			if (cnt % 2 == 0) Answer -= num;
		}
		
		// 출력
		System.out.println(Answer);
	}
	static long GCD(long A, long B) {
		// 최대공약수를 리턴하는 함수
		if (B == 0) return A;
		return GCD(B, A % B);
	}
	static long LCM(long A, long B) {
		// 최소공배수를 리턴하는 함수
		return (A / GCD(A, B)) * B;
	}
}

File: 문제풀이/chap5-4/prob5-4-5.py
# 최대공약수를 리턴하는 함수
def GCD(A, B):
	while A >= 1 and B >= 1:
		if A < B:
			B = B % A  # A < B라면, 큰 수를 b로 변경합니다.
		else:
			A = A % B  # A >= B라면 큰 수를 a로 변경합니다.
	if A >= 1:
		return A
	return B

# 최소공배수를 리턴하는 함수
def LCM(A, B):
	return int(A / GCD(A, B)) * B

# 입력
N, K = map(int, input().split())
V = list(map(int, input().split()))

# 비트 전체 탐색
Answer = 0
for i in range(1, 1 << K):
	cnt = 0
	lcm = 1
	for j in range(K):
		if (i & (1 << j)) != 0:
			cnt += 1
			lcm = LCM(lcm, V[j])
			
	# num은 N 이하 중에서 선택된 숫자 모두의 배수인 것의 개수
	num = N // lcm
	if cnt % 2 == 1:
		Answer += num
	else:
		Answer -= num

# 출력
print(Answer)